source/opt/constructs.h

```diff
@@ -0,0 +1,305 @@
+  // Returns the vector of line-related debug instructions attached to this
+  // instruction and the caller can directly modify them.
+  std::vector<Instruction>& dbg_line_insts() { return dbg_line_insts_; }
+
+  // Gets the number of logical operands.
+  uint32_t NumOperands() const {
+    return static_cast<uint32_t>(operands_.size());
+  }
+  // Gets the number of SPIR-V words occupied by all logical operands.
+  uint32_t NumOperandWords() const {
+    return NumInOperandWords() + TypeResultIdCount();
+  }
+  // Gets the |index|-th logical operand.
+  const Operand& GetOperand(uint32_t index) const;
+  // Gets the |index|-th logical operand as a single SPIR-V word. This will only
+  // consider the first word in the logical operand, even which has more than
+  // one words.
+  uint32_t GetSingleWordOperand(uint32_t index) const;
+  // Sets the |index|-th operand's data to the given |data|.
+  void SetOperand(uint32_t index, std::vector<uint32_t>&& data);
+
+  // The following methods are similar to the above, but are for in operands.
+  uint32_t NumInOperands() const {
+    return static_cast<uint32_t>(operands_.size() - TypeResultIdCount());
+  }
+  uint32_t NumInOperandWords() const;
+  const Operand& GetInOperand(uint32_t index) const {
+    return GetOperand(index + TypeResultIdCount());
+  }
+  uint32_t GetSingleWordInOperand(uint32_t index) const {
+    return GetSingleWordOperand(index + TypeResultIdCount());
+  }
+
+  // Returns true if this instruction is OpNop.
+  bool IsNop() const {
+    return opcode_ == SpvOpNop && type_id_ == 0 && result_id_ == 0 &&
+           operands_.empty();
+  }
+  // Turns this instruction to OpNop.
+  void ToNop() {
+    opcode_ = SpvOpNop;
+    type_id_ = result_id_ = 0;
+    operands_.clear();
+  }
+
+  // Runs the given function |f| on this instruction and preceding debug
+  // instructions.
+  void ForEachInst(const std::function<void(Instruction*)>& f);
+
+  // Pushes the binary segments for this instruction into the back of *|binary|.
+  // If |keep_nop| is true and this is a OpNop, do nothing.
+  void ToBinary(std::vector<uint32_t>* binary, bool keep_nop) const;
+
+ private:
+  // Returns the toal count of result type id and result id.
+  uint32_t TypeResultIdCount() const {
+    return (type_id_ != 0) + (result_id_ != 0);
+  }
+
+  SpvOp opcode_;        // Opcode
+  uint32_t type_id_;    // Result type id
+  uint32_t result_id_;  // Result id
+  // All logical operands, including result type id and result id.
+  std::vector<Operand> operands_;
+  // Opline and OpNoLine instructions preceding this instruction. Note that for
+  // Instructions representing OpLine or OpNonLine itself, this field should be
+  // empty.
+  std::vector<Instruction> dbg_line_insts_;
+};
+
+// A SPIR-V basic block.
+class BasicBlock {
+ public:
+  // Creates a basic block with the given enclosing |function| and starting
+  // |label|.
+  BasicBlock(Function* function, Instruction&& label)
```

Comment:
Again, since the label is an rvalue reference, should I assume it's implied that this constructor takes ownership of the label instruction?
Should we say so in the blurb?  I'm leaning toward yes.